turbopack/crates/turbo-persistence/src/static_sorted_file_builder.rs

```diff
@@ -158,114 +154,134 @@ pub fn write_static_stored_file<E: Entry>(
+            if range.start == 0 && range.end == key_len {
+                entry.write_key_to(buf);
+            } else {
+                let mut temp = Vec::with_capacity(key_len);
+                entry.write_key_to(&mut temp);
+                debug_assert!(temp.len() == key_len);
+
+                buf.extend_from_slice(&temp[range]);
+            }
+        },
+        buffer,
+    )
+    .context("Key dictionary creation failed")
+    .unwrap())
+}
+
 /// Computes compression dictionaries from values of all entries
 #[tracing::instrument(level = "trace", skip(entries))]
 fn compute_value_compression_dictionary<E: Entry>(
     entries: &[E],
     total_value_size: usize,
     buffer: &mut Vec<u8>,
 ) -> Result<Vec<u8>> {
-    if total_value_size < MIN_VALUE_COMPRESSION_SAMPLES_SIZE {
-        return Ok(Vec::new());
-    }
-    let value_compression_samples_size = min(VALUE_COMPRESSION_SAMPLES_SIZE, total_value_size / 16);
-    let mut sample_sizes = Vec::new();
-
-    // Limit the number of iterations to avoid infinite loops
-    let max_iterations = total_value_size / COMPRESSION_DICTIONARY_SAMPLE_PER_ENTRY * 2;
-    for i in 0..max_iterations {
-        let entry = &entries[i % entries.len()];
-        let remaining = value_compression_samples_size - buffer.len();
-        if remaining < MIN_COMPRESSION_DICTIONARY_SAMPLE_PER_ENTRY {
-            break;
-        }
-        if let EntryValue::Small { value } | EntryValue::Medium { value } = entry.value() {
-            let len = value.len();
-            if len >= MIN_COMPRESSION_DICTIONARY_SAMPLE_PER_ENTRY {
-                let used_len = min(remaining, COMPRESSION_DICTIONARY_SAMPLE_PER_ENTRY);
-                if len <= used_len {
-                    sample_sizes.push(len);
-                    buffer.extend_from_slice(value);
-                } else {
-                    sample_sizes.push(used_len);
-                    let p = buffer.len() % (len - used_len);
-                    buffer.extend_from_slice(&value[p..p + used_len]);
-                };
+    compute_dictionary(
+        entries,
+        total_value_size,
+        VALUE_COMPRESSION_SAMPLES_SIZE,
+        VALUE_COMPRESSION_DICTIONARY_SIZE,
+        |e| {
+            if let EntryValue::Small { value } | EntryValue::Medium { value } = e.value() {
+                value.len()
+            } else {
+                0
             }
-        }
-    }
-    debug_assert!(buffer.len() == sample_sizes.iter().sum::<usize>());
-    let result = if buffer.len() > MIN_VALUE_COMPRESSION_SAMPLES_SIZE && sample_sizes.len() > 5 {
-        zstd::dict::from_continuous(buffer, &sample_sizes, VALUE_COMPRESSION_DICTIONARY_SIZE)
-            .context("Value dictionary creation failed")?
-    } else {
-        Vec::new()
-    };
-    buffer.clear();
-    Ok(result)
+        },
+        |entry, range, buf| {
+            if let EntryValue::Small { value } | EntryValue::Medium { value } = entry.value() {
```

Comment:
Should we decompress compressed blocks here during compaction to get a better compression dictionary?